sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,479 @@
+   * If isNumeric is true, then it is a numeric value.  Otherwise, it is a Literal value.
+   */
+  def numericLiteralToBigDecimal(
+       literal: Any,
+       dataType: DataType,
+       isNumeric: Boolean = false)
+    : BigDecimal = {
+    dataType match {
+      case _: IntegralType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[Long])
+        else BigDecimal(literal.asInstanceOf[Literal].value.asInstanceOf[Long])
+      case _: FractionalType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[Double])
+        else BigDecimal(literal.asInstanceOf[Literal].value.asInstanceOf[Double])
+      case DateType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[BigInt])
+        else {
+          val dateLiteral = DateTimeUtils.stringToDate(
+            literal.asInstanceOf[Literal].value.asInstanceOf[UTF8String])
+          BigDecimal(dateLiteral.asInstanceOf[BigInt])
+        }
+      case TimestampType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[BigInt])
+        else {
+          val tsLiteral = DateTimeUtils.stringToTimestamp(
+            literal.asInstanceOf[Literal].value.asInstanceOf[UTF8String])
+          BigDecimal(tsLiteral.asInstanceOf[BigInt])
+        }
+    }
+  }
+
+  /** This method evaluates the equality predicate for all data types. */
+  def evaluateEqualTo(
+      op: BinaryComparison,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+
+    /**
+     * decide if the value is in [min, max] of the column.
+     * We currently don't store min/max for binary/string type.
+     * Hence, we assume it is in boundary for binary/string type.
+     */
+    val inBoundary: Boolean = attrRef.dataType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, attrRef.dataType).asInstanceOf[NumericRange]
+        val lit = numericLiteralToBigDecimal(literal, attrRef.dataType)
+        (lit >= statsRange.min) && (lit <= statsRange.max)
+
+      case _ => true  /** for String/Binary type */
+    }
+
+    val percent: Double =
+      if (inBoundary) {
+
+        if (update) {
+          /**
+           * We update ColumnStat structure after apply this equality predicate.
+           * Set distinctCount to 1.  Set nullCount to 0.
+           */
+          val newStats = attrRef.dataType match {
+            case _: NumericType | DateType | TimestampType =>
+              val newValue = Some(literal.value)
+              aColStat.copy(distinctCount = 1, min = newValue,
+                max = newValue, nullCount = 0)
+            case _ => aColStat.copy(distinctCount = 1, nullCount = 0)
+          }
+          mutableColStats += (attrRef.exprId -> newStats)
+        }
+
+        1.0 / ndv.toDouble
+      } else {
+        0.0
+      }
+
+    percent
+  }
+
+  def evaluateInSet(

```

Comment:
Yes, the return value is a double value showing the percentage of rows meeting a given condition. Also I will add comments for this method in JavaDoc style.